WireUI/Sources/WireUIFoundation/Main/MainCoordinator.swift

```diff
@@ -0,0 +1,364 @@
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+import WireFoundation
+
+// swiftlint:disable opening_brace
+
+/// Manages the main navigation and the layout changes of the application after a successful login.
+///
+/// The MainCoordinator class is the central controller for the app's navigation and layout management.
+/// It receives references to ``MainTabBarControllerProtocol`` and ``MainSplitViewControllerProtocol``
+/// conforming instances and is responsible for managing transitions between different split layout states (collapsed and expanded)
+/// as well as handling navigation logic.
+///
+/// Both, the split view controller as well as the tab view controller actually install `UINavigationController`
+/// instances and then put or remove the content view controllers into/from `viewControllers` array.
+
+@MainActor
+public final class MainCoordinator<
+
+    SplitViewController: MainSplitViewControllerProtocol,
+    TabBarController: MainTabBarControllerProtocol,
+    NewConversationBuilder: MainCoordinatorInjectingViewControllerBuilder,
+    SelfProfileBuilder: MainCoordinatorInjectingViewControllerBuilder
+
+>: NSObject, MainCoordinatorProtocol, UISplitViewControllerDelegate, UITabBarControllerDelegate where
+
+    SplitViewController.Sidebar: MainSidebarProtocol,
+    SplitViewController.ConversationList == TabBarController.ConversationList,
+    SplitViewController.Archive == TabBarController.Archive,
+    SplitViewController.Settings == TabBarController.Settings
+{
+    // swiftlint:enable opening_brace
+
+    // MARK: - Private Properties
+
+    private weak var splitViewController: SplitViewController!
+    private weak var tabBarController: TabBarController!
+
+    private let newConversationBuilder: NewConversationBuilder
+    private weak var newConversation: UIViewController?
+
+    private var selfProfileBuilder: SelfProfileBuilder
+    private weak var selfProfile: UIViewController?
+
+    private var mainSplitViewState: MainSplitViewState = .expanded
+
+    // MARK: - Private Helpers
+
+    private var sidebar: SplitViewController.Sidebar! {
+        splitViewController.sidebar
+    }
+
+    private var conversationList: TabBarController.ConversationList! {
+        switch mainSplitViewState {
+        case .collapsed: tabBarController.conversations?.conversationList
+        case .expanded: splitViewController.conversationList
+        }
+    }
+
+    private var archive: TabBarController.Archive! {
+        switch mainSplitViewState {
+        case .collapsed: tabBarController.archive
+        case .expanded: splitViewController.archive
+        }
+    }
+
+    private var settings: TabBarController.Settings! {
+        switch mainSplitViewState {
+        case .collapsed: tabBarController.settings
+        case .expanded: splitViewController.settings
+        }
+    }
+
+    // MARK: - Life Cycle
+
+    public init(
```

Comment:
In my opinion we should throw errors only for issues which could go wrong due to specific circumstances at runtime (network, memory, user input or action), unless we're writing a library for somebody else.

Initializing the MainCoordinator in a wrong way is a mistake on the developer's side. For those case I usually want fail hard so that the issue is discovered early.